### ISSUE TITLE
after run pyinstaller -D -w xxx.py

### ISSUE THREAD

**Issue by cqray1990** [pyinstaller/pyinstaller#7697]
### Stacktrace / full error message
after run  pyinstaller -D -w xxx.py and exectue the exe, raisr error as follows:

the code is:
```
def set_paddle_lib_path():
    site_dirs = (
        site.getsitepackages()
        if hasattr(site, 'getsitepackages')
        else [x for x in sys.path if 'site-packages' in x]
    )
    for site_dir in site_dirs:
        lib_dir = os.path.sep.join([site_dir, 'paddle', 'libs'])
        if os.path.exists(lib_dir):
            _set_paddle_lib_path(lib_dir)
            set_paddle_custom_device_lib_path(
                os.path.sep.join([lib_dir, '..', '..', 'paddle_custom_device'])
            )
            return
    if getattr(sys, 'frozen', False):
        basedir = sys._MEIPASS
    else:
        basedir = os.path.dirname(os.path.abspath(__file__))
    print(basedir)
    if hasattr(site, 'USER_SITE'):
        #lib_dir = os.path.sep.join([site.USER_SITE, 'paddle', 'libs'])
        lib_dir = os.path.join(basedir, 'paddle', 'libs')
        if os.path.exists(lib_dir):
            _set_paddle_lib_path(lib_dir)
            set_paddle_custom_device_lib_path(
                os.path.sep.join([lib_dir, '..', '..', 'paddle_custom_device'])
            )
```

  File "paddle/fluid/core.py", line 385, in set_paddle_lib_path
    basedir = sys._MEIPASS
TypeError: sequence item 0: expected str instance, NoneType found
[31062] Failed to execute script 'extract' due to unhandled exception!

**Comment by rokm** [comment]
Looks like [`paddle` is assuming that if `site` has `USER_SITE` attribute, it cannot be `None`](https://github.com/PaddlePaddle/Paddle/blob/v2.4.2/python/paddle/fluid/core.py#L352-L373). But if you run python with site-disabled (`-S`) switch, it is `None` (of course, with `site` disabled, you cannot import `paddle` from `site-packages` in first place, so for unfrozen code, this is not really a problem...). PyInstaller-frozen application runs with equivalent of `-S` switch, so `site` is disabled, and `site.USER_SITE` is `None`. Hence the error.

You seem to have already tried to work around this, and I think that should work.

The following is a bit cleaner, and gets me past that error with `paddlepaddle==2.4.2` (the CPU version) with simple `import paddle` program:

```python
# set paddle lib path
def set_paddle_lib_path():
    site_dirs = (
        site.getsitepackages()
        if hasattr(site, 'getsitepackages')
        else [x for x in sys.path if 'site-packages' in x]
    )
    for site_dir in site_dirs:
        lib_dir = os.path.sep.join([site_dir, 'paddle', 'libs'])
        if os.path.exists(lib_dir):
            _set_paddle_lib_path(lib_dir)
            set_paddle_custom_device_lib_path(
                os.path.sep.join([lib_dir, '..', '..', 'paddle-plugins'])
            )
            return
    if hasattr(site, 'USER_SITE') and site.USER_SITE is not None:
        lib_dir = os.path.sep.join([site.USER_SITE, 'paddle', 'libs'])
        if os.path.exists(lib_dir):
            _set_paddle_lib_path(lib_dir)
            set_paddle_custom_device_lib_path(
                os.path.sep.join([lib_dir, '..', '..', 'paddle-plugins'])
            )
            return
    if hasattr(sys, '_MEIPASS'):
        lib_dir = os.path.sep.join([sys._MEIPASS, 'paddle', 'libs'])
        if os.path.exists(lib_dir):
            _set_paddle_lib_path(lib_dir)
            set_paddle_custom_device_lib_path(
                os.path.sep.join([lib_dir, '..', '..', 'paddle-plugins'])
            )
```

---

Then the next error is due to the import in auto-generated `paddle/fluid/proto/pass_desc_pb2.py`, which should be changed from

```
import framework_pb2 as framework__pb2
```

into


```
from . import framework_pb2 as framework__pb2
```

---

After that is fixed, the frozen application [starts respawning itself due to use of `subprocess` with `sys.executable`](https://github.com/PaddlePaddle/Paddle/blob/0e92adceae06b6b7463f2dc7790ffb0601730009/python/paddle/dataset/image.py#L38-L64), which is incompatible with PyInstaller (see also #5902). Brénainn has submitted a fix for that to `paddlepaddle` (PaddlePaddle/Paddle#33543), but it was not merged.

---

Also note that even if you fix `set_paddle_lib_path`, the `paddle/libs` directory will not be collected unless you explicitly collect the shared libraries from it (the ones that were discovered via binary analysis will end up in sys._MEIPASS instead). Same goes for that `paddle-plugins` directory (and if `paddle` expects the plugins to be collected as source files, you'll need to accommodate that).

---

So in short, I consider `paddlepaddle` to be pretty much incompatible with and unsupported by PyInstaller, so if you want to use it, do so at your own risk, and expect that you will need to work around various issues on your own.
